tls/s2n_recv.c

```diff
@@ -134,6 +134,15 @@ ssize_t s2n_recv_impl(struct s2n_connection *conn, void *buf, ssize_t size_signe
         return 0;
     }
 
+    /*
+     * Set the `blocked` status to BLOCKED_ON_READ by default
+     *
+     * The only case in which it should be updated is on a successful read into the provided buffer.
+     *
+     * Unfortunately, the current `blocked` behavior has become ossified by buggy applications that ignore
+     * error types and only read `blocked`. As such, it's very important to avoid changing how this value is updated
+     * as it could break applications.
+     */
     *blocked = S2N_BLOCKED_ON_READ;
```

Comment:
When `s2n_recv()` is called and the socket is closed, the [`conn->read_closed` flag is set](https://github.com/aws/s2n-tls/blob/17eae9d45cb7b588fd15ad06441153c7b16b8ba5/tls/s2n_recv.c#L45). The next time `s2n_recv()` is called, we [check this flag](https://github.com/aws/s2n-tls/blob/17eae9d45cb7b588fd15ad06441153c7b16b8ba5/tls/s2n_recv.c#L119) via `s2n_connection_check_io_status`. If the socket was closed but a close notify wasn't received, we [bail with `S2N_ERR_CLOSED`](https://github.com/aws/s2n-tls/blob/17eae9d45cb7b588fd15ad06441153c7b16b8ba5/tls/s2n_recv.c#L132). However, `blocked` isn't updated before bailing in this case.

For applications that behave as described in https://github.com/aws/s2n-tls/pull/4127#discussion_r1282361066, I worry that this change introduces a possibility of applications getting stuck in an infinite loop. If `s2n_recv()` is called by the application, we could read some bytes and then detect that the socket is closed, setting the `conn->read_closed` flag. After this change, since we now read at least one byte, we set `blocked` to `S2N_NOT_BLOCKED`. This will cause the application to loop again to read more data from `s2n_recv()`, as intended. However, this time `s2n_recv()` will bail with `S2N_ERR_CLOSED`, since the socket was closed and a close notify wasn't received. Since `blocked` isn't updated with `S2N_BLOCKED_ON_READ` before bailing, applications that don't check for error codes will continue calling `s2n_recv()`, leading to infinite looping.

To fix this we could move the `*blocked = S2N_BLOCKED_ON_READ;` line to the top of the function, which would set `blocked` before bailing with `S2N_ERR_CLOSED`.

Reply:
Yeah moving that to the top seems reasonable to me